test/AutoDiff/leakchecking.swift

```diff
@@ -208,6 +208,116 @@ LeakCheckingTests.testWithLeakChecking("ClassMethods") {
   expectEqual((3, 3), classValueWithGradient(SubOverrideCustomDerivatives()))
 }
 
+protocol TF_508_Proto {
+  associatedtype Scalar
+}
+extension TF_508_Proto where Scalar : FloatingPoint {
+  @differentiable(
+    jvp: jvpAdd, vjp: vjpAdd
+    where Self : Differentiable, Scalar : Differentiable,
+          // Conformance requirement with dependent member type.
+          Self.TangentVector : TF_508_Proto
+  )
+  static func +(lhs: Self, rhs: Self) -> Self {
+    return lhs
+  }
+
+  @differentiable(
+    jvp: jvpSubtract, vjp: vjpSubtract
+    where Self : Differentiable, Scalar : Differentiable,
+          // Same-type requirement with dependent member type.
+          Self.TangentVector == TF_508_Struct<Float>
+  )
+  func subtract(_ other: Self) -> Self {
+    return self
+  }
+}
+extension TF_508_Proto where Self : Differentiable,
+                             Scalar : FloatingPoint & Differentiable,
+                             Self.TangentVector : TF_508_Proto {
+  static func jvpAdd(lhs: Self, rhs: Self)
+    -> (Self, (TangentVector, TangentVector) -> TangentVector) {
+    return (lhs, { (dlhs, drhs) in dlhs + drhs })
+  }
+  static func vjpAdd(lhs: Self, rhs: Self)
+    -> (Self, (TangentVector) -> (TangentVector, TangentVector)) {
+    return (lhs, { v in (v, v) })
+  }
+}
+extension TF_508_Proto where Self : Differentiable,
+                             Scalar : FloatingPoint & Differentiable,
+                             Self.TangentVector == TF_508_Struct<Float> {
+  func jvpSubtract(lhs: Self)
+    -> (Self, (TangentVector, TangentVector) -> TangentVector) {
+    return (lhs, { dself, dlhs in dself - dlhs })
+  }
+  func vjpSubtract(lhs: Self)
+    -> (Self, (TangentVector) -> (TangentVector, TangentVector)) {
+    return (lhs, { v in (v, v) })
+  }
+}
+
+struct TF_508_Struct<Scalar : AdditiveArithmetic>
+  : TF_508_Proto, AdditiveArithmetic {}
+extension TF_508_Struct : Differentiable where Scalar : Differentiable {
+  typealias TangentVector = TF_508_Struct
+}
+
+// Test leaks regarding `SILGenFunction::getOrCreateAutoDiffLinearMapThunk`.
+LeakCheckingTests.testWithLeakChecking("LinearMapSILGenThunks") {
+  func testLinearMapSILGenThunks() {
+    let x = TF_508_Struct<Float>()
+    // Test conformance requirement with dependent member type.
+    _ = pullback(at: x, in: { (x: TF_508_Struct<Float>) -> TF_508_Struct<Float> in
+      return x + x
+    })
+    // Test same-type requirement with dependent member type.
+    _ = pullback(at: x, in: { (x: TF_508_Struct<Float>) -> TF_508_Struct<Float> in
+      return x.subtract(x)
+    })
+  }
+  testLinearMapSILGenThunks()
+}
+
+LeakCheckingTests.testWithLeakChecking("ParameterConventionMismatchLeakChecking") {
```

Comment:
Note: here are leak checking tests for [TF-667](https://bugs.swift.org/projects/TF/issues/TF-667).